File: web/public/js/paginas/composicion.js
/* Composición del panel y universo de referencia (R2.2, R2.3 + P1).

   Dos cosas en la misma pantalla, porque no se entienden por separado: la
   composición observada del panel y el universo con el que se la compara.

   Lo que la pantalla tiene que dejar claro es la diferencia entre «no hay
   brecha» y «no se puede calcular la brecha». Sin objetivo cargado, la
   composición es descriptiva y la brecha aparece explícitamente como no
   disponible: una columna vacía se leería como «está todo bien».
*/

import * as api from '../api.js';
import {
  $, $$, esc, encabezado, vacio, cargando, toast, modal, cerrarModal,
  leerFormulario, alerta, confirmar,
} from '../ui.js';

let contexto = {};
let panelActual = null;
let cruceActual = null;

const DIMENSIONES = {
  sexo: 'Sexo',
  tramo_etario: 'Tramo etario',
  localidad: 'Localidad',
};

const TRAMOS = ['<18', '18-24', '25-34', '35-44', '45-54', '55-64', '65+'];

const pct = (n) => (n == null ? '—' : `${(n * 100).toFixed(1)} %`);

export async function render(main, ctx) {
  contexto = ctx;
  const { items: paneles } = await api.paneles.listar();
  const activos = paneles.filter((p) => p.estado === 'activo');
  if (!activos.length) {
    main.innerHTML = encabezado('Composición', 'del panel', '')
      + vacio('No hay paneles activos todavía.', '📋');
    return;
  }
  panelActual = activos.some((p) => p.id === panelActual) ? panelActual : activos[0].id;

  main.innerHTML = encabezado('Composición', 'y brecha',
    'Cuánto se parece el panel al universo que pretende representar.') + `
    <div class="card"><div class="card-body">
      <div class="form-row" style="margin:0">
        <div class="form-group" style="margin:0"><label>Panel</label>
          <select class="fselect" id="panel">${activos.map((p) =>
            `<option value="${p.id}" ${p.id === panelActual ? 'selected' : ''}>${esc(p.nombre)}</option>`
          ).join('')}</select></div>
        <div class="form-group" style="margin:0"><label>Cruce de dos dimensiones</label>
          <select class="fselect" id="cruce">
            <option value="">Sin cruce</option>
            <option value="sexo,tramo_etario">Sexo × Tramo etario</option>
            <option value="sexo,localidad">Sexo × Localidad</option>
            <option value="tramo_etario,localidad">Tramo etario × Localidad</option>
          </select></div>
      </div>
    </div></div>
    <div id="cuerpo">${cargando()}</div>`;

  $('#panel').onchange = (e) => { panelActual = Number(e.target.value); cargar(); };
  $('#cruce').value = cruceActual ? cruceActual.join(',') : '';
  $('#cruce').onchange = (e) => {
    cruceActual = e.target.value ? e.target.value.split(',') : null;
    cargar();
  };
  await cargar();
}

async function cargar() {
  const caja = $('#cuerpo');
  caja.innerHTML = cargando();
  try {
    const salida = await api.composicion.ver(panelActual, { cruce: cruceActual });
    caja.innerHTML = pintar(salida);
    $('#cargar-objetivo').onclick = () => abrirObjetivo(salida);
    $$('[data-borrar-objetivo]', caja).forEach((b) => {
      b.onclick = () => borrarObjetivo(b.dataset.borrarObjetivo);
    });
  } catch (error) {
    caja.innerHTML = alerta(error.message);
  }
}

function pintar(salida) {
  const puedeGestionar = ['admin', 'operaciones'].includes(contexto.actor?.rol);
  return `
    <div class="stat-grid">
      <div class="stat s-total"><label>Miembros activos</label><strong>${salida.miembros}</strong></div>
      <div class="stat ${salida.objetivo_cargado ? 's-ok' : 's-warn'}">
        <label>Universo de referencia</label>
        <strong>${salida.objetivo_cargado ? 'cargado' : 'sin cargar'}</strong></div>
    </div>

    ${salida.objetivo_cargado ? '' : `<div class="alert alert-warn">
      Sin universo de referencia cargado la composición es <strong>solo
      descriptiva</strong>: la brecha no se puede calcular, no es que sea
      cero. ${puedeGestionar ? 'Cargá el objetivo para poder compararla.' : ''}
    </div>`}

    <div class="toolbar" style="margin-bottom:1rem">
      <button class="btn btn-orange" id="cargar-objetivo"
        ${puedeGestionar ? '' : 'disabled title="Hace falta rol de operaciones o admin"'}>
        ${salida.objetivo_cargado ? 'Editar universo de referencia' : 'Cargar universo de referencia'}
      </button>
    </div>

    ${salida.dimensiones.map((d) => pintarDimension(d, salida.miembros, puedeGestionar)).join('')}
    ${salida.cruce ? pintarCruce(salida.cruce) : ''}`;
}

function pintarDimension(dimension, miembros, puedeGestionar) {
  const hay = dimension.brecha_disponible;
  return `<div class="card">
    <div class="card-header">
      <span class="card-header-title">${esc(DIMENSIONES[dimension.dimension] || dimension.dimension)}</span>
      <div class="toolbar">
        ${hay ? `<span class="small muted">disimilitud ${pct(dimension.disimilitud)}</span>
          ${puedeGestionar ? `<button class="btn btn-outline btn-sm btn-del"
            data-borrar-objetivo="${dimension.dimension}">Quitar objetivo</button>` : ''}`
        : '<span class="badge badge-off">brecha no disponible</span>'}
      </div>
    </div>
    <div class="card-body tight">
      ${hay ? '' : `<div class="small muted" style="padding:.5rem 1.5rem 0">
        ${esc(dimension.motivo_sin_brecha)}</div>`}
      <div class="table-wrap"><table>
        <thead><tr>
          <th>Categoría</th><th>Observados</th><th>% observado</th>
          <th>% objetivo</th><th>Brecha</th><th>Faltan / sobran</th>
        </tr></thead>
        <tbody>${dimension.categorias.map((c) => `<tr>
          <td class="td-strong">${esc(c.categoria)}</td>
          <td class="mono">${c.observados}</td>
          <td><div class="barra ${hay && Math.abs(c.brecha ?? 0) < 0.03 ? 'ok' : ''}">
                <span style="width:${Math.round((c.proporcion_observada || 0) * 100)}%"></span></div>
              <div class="small mono">${pct(c.proporcion_observada)}</div></td>
          <td class="mono">${pct(c.proporcion_objetivo)}</td>
          <td class="mono ${c.brecha == null ? 'muted'
            : (c.brecha < 0 ? 'brecha-falta' : c.brecha > 0 ? 'brecha-sobra' : '')}">
            ${c.brecha == null ? '—' : `${c.brecha > 0 ? '+' : ''}${(c.brecha * 100).toFixed(1)} pp`}</td>
          <td class="small">${c.faltan ? `faltan <strong>${c.faltan}</strong>`
            : c.sobran ? `sobran ${c.sobran}` : (hay ? 'calza' : '—')}</td>
        </tr>`).join('')}</tbody>
      </table></div>
    </div>
  </div>`;
}

function pintarCruce(cruce) {
  const [a, b] = cruce.dimensiones;
  const filas = [...new Set(cruce.celdas.map((c) => c[a]))].sort();
  const columnas = [...new Set(cruce.celdas.map((c) => c[b]))].sort();
  const valor = {};
  cruce.celdas.forEach((c) => { valor[`${c[a]}|${c[b]}`] = c.observados; });

  return `<div class="card">
    <div class="card-header">
      <span class="card-header-title">
        ${esc(DIMENSIONES[a] || a)} × ${esc(DIMENSIONES[b] || b)}</span>
      <span class="badge badge-off">descriptivo</span>
    </div>
    <div class="card-body tight">
      <div class="small muted" style="padding:.5rem 1.5rem 0">
        ${esc(cruce.motivo_sin_brecha)} Las celdas vacías son los huecos que
        las marginales esconden.
      </div>
      <div class="table-wrap"><table>
        <thead><tr><th>${esc(DIMENSIONES[a] || a)}</th>
          ${columnas.map((c) => `<th>${esc(c)}</th>`).join('')}<th>Total</th></tr></thead>
        <tbody>${filas.map((f) => {
          const total = columnas.reduce((s, c) => s + (valor[`${f}|${c}`] || 0), 0);
          return `<tr><td class="td-strong">${esc(f)}</td>
            ${columnas.map((c) => {
              const n = valor[`${f}|${c}`] || 0;
              return `<td class="mono ${n ? '' : 'td-muted'}">${n || '·'}</td>`;
            }).join('')}
            <td class="mono td-strong">${total}</td></tr>`;
        }).join('')}</tbody>
      </table></div>
    </div>
  </div>`;
}

/* ── Carga del universo de referencia ───────────────────────────── */

function abrirObjetivo(salida) {
  const actuales = {};
  salida.dimensiones.forEach((d) => {
    d.categorias.forEach((c) => {
      if (c.proporcion_objetivo != null) {
        actuales[`${d.dimension}|${c.categoria}`] = c.proporcion_objetivo;
      }
    });
  });
  const observadas = {};
  salida.dimensiones.forEach((d) => {
    observadas[d.dimension] = d.categorias.map((c) => c.categoria);
  });

  const dimension = 'sexo';
  const caja = modal({
    titulo: 'Universo de referencia',
    ancho: '680px',
    cuerpo: `
      <div class="alert alert-info">
        Se carga <strong>una dimensión a la vez</strong>, en proporciones que
        sumen 1. Una dimensión que suma 0,8 no es un universo incompleto:
        haría que todas sus brechas estén mal, así que se rechaza.
      </div>
      <div class="form-group"><label>Dimensión</label>
        <select class="fselect" id="obj-dimension">
          ${Object.entries(DIMENSIONES).map(([k, v]) =>
            `<option value="${k}">${esc(v)}</option>`).join('')}
        </select></div>
      <div id="obj-categorias"></div>
      <div class="fila-criterio"><div>Suma de las proporciones</div>
        <span id="obj-suma" class="mono">0.000</span></div>
      <div class="toolbar">
        <button class="btn btn-outline btn-sm" id="obj-agregar">+ Categoría</button>
        <button class="btn btn-outline btn-sm" id="obj-desde-panel">
          Precargar con lo observado</button>
      </div>`,
    acciones: [
      { texto: 'Cancelar', clase: 'btn-outline', onClick: cerrarModal },
      { texto: 'Guardar', clase: 'btn-orange', onClick: guardarObjetivo },
    ],
  });

  const pintarCategorias = (dim) => {
    const propias = Object.entries(actuales)
      .filter(([clave]) => clave.startsWith(`${dim}|`))
      .map(([clave, valor]) => [clave.split('|')[1], valor]);
    const sugeridas = dim === 'tramo_etario' ? TRAMOS
      : dim === 'sexo' ? ['F', 'M']
      : (observadas[dim] || []).filter((c) => c !== '(sin dato)');
    const lista = propias.length ? propias
      : sugeridas.map((c) => [c, 0]);
    $('#obj-categorias', caja).innerHTML = lista.map(([cat, prop]) => fila(cat, prop)).join('');
    enganchar();
  };

  const fila = (categoria = '', proporcion = 0) => `
    <div class="pregunta-fila" data-fila>
      <input type="text" data-cat placeholder="Categoría" value="${esc(categoria)}" />
      <input type="number" data-prop step="0.01" min="0" max="1"
             placeholder="0.50" value="${proporcion || ''}" />
      <button class="btn btn-outline btn-sm btn-del" data-quitar>×</button>
    </div>`;

  const recalcular = () => {
    const suma = $$('[data-prop]', caja)
      .reduce((s, i) => s + (Number(i.value) || 0), 0);
    const nodo = $('#obj-suma', caja);
    nodo.textContent = suma.toFixed(3);
    nodo.className = `mono ${Math.abs(suma - 1) <= 0.005 ? '' : 'brecha-falta'}`;
  };

  const enganchar = () => {
    $$('[data-prop]', caja).forEach((i) => { i.oninput = recalcular; });
    $$('[data-quitar]', caja).forEach((b) => {
      b.onclick = () => { b.closest('[data-fila]').remove(); recalcular(); };
    });
    recalcular();
  };

  $('#obj-dimension', caja).onchange = (e) => pintarCategorias(e.target.value);
  $('#obj-agregar', caja).onclick = () => {
    $('#obj-categorias', caja).insertAdjacentHTML('beforeend', fila());
    enganchar();
  };
  $('#obj-desde-panel', caja).onclick = () => {
    const dim = $('#obj-dimension', caja).value;
    const d = salida.dimensiones.find((x) => x.dimension === dim);
    if (!d || !salida.miembros) { toast('No hay observados para precargar.', 'err'); return; }
    $('#obj-categorias', caja).innerHTML = d.categorias
      .filter((c) => c.categoria !== '(sin dato)')
      .map((c) => fila(c.categoria, Number(c.proporcion_observada.toFixed(2)))).join('');
    enganchar();
    toast('Precargado con la composición actual. Ajustalo al universo real.', 'ok');
  };

  pintarCategorias(dimension);
}

async function guardarObjetivo(caja) {
  const dimension = $('#obj-dimension', caja).value;
  const objetivos = $$('[data-fila]', caja).map((fila) => ({
    dimension,
    categoria: $('[data-cat]', fila).value.trim(),
    proporcion: Number($('[data-prop]', fila).value),
  })).filter((o) => o.categoria);

  if (!objetivos.length) { toast('Cargá al menos una categoría.', 'err'); return; }
  try {
    await api.composicion.cargarObjetivo(panelActual, objetivos);
    cerrarModal();
    toast(`Universo de referencia de «${DIMENSIONES[dimension]}» guardado.`, 'ok');
    await cargar();
  } catch (error) {
    toast(error.detalle?.suma != null
      ? `Las proporciones suman ${error.detalle.suma} y tienen que sumar 1.`
      : error.message, 'err');
  }
}

async function borrarObjetivo(dimension) {
  const ok = await confirmar({
    titulo: `Quitar el objetivo de ${DIMENSIONES[dimension] || dimension}`,
    cuerpo: `<p>La composición de esa dimensión vuelve a ser descriptiva y la
      brecha deja de estar disponible.</p>`,
    textoOk: 'Quitar',
  });
  if (!ok) return;
  try {
    await api.composicion.borrarObjetivo(panelActual, dimension);
    toast('Objetivo quitado.', 'ok');
    await cargar();
  } catch (error) {
    toast(error.message, 'err');
  }
}

File: functions/panel_api/composicion.py
"""R2.2 y R2.3 — universo de referencia, composición observada y brecha.

Un panel no se juzga por su tamaño sino por su parecido con el universo que
pretende representar. Este módulo hace las dos mitades de eso:

* **R2.2 — cargar el objetivo.** Un universo de referencia es un conjunto de
  proporciones por dimensión y categoría: sexo 52 % F / 48 % M, tramo etario
  18-24 14 %, 25-34 19 %, etc. Se valida que las proporciones de **cada
  dimensión sumen 1**; una dimensión que suma 0,8 o 1,3 no es un universo,
  es un error de carga, y aceptarlo haría que todas las brechas de esa
  dimensión mientan.

* **R2.3 — describir y comparar.** Con objetivo cargado se devuelve la
  composición observada junto a la brecha. Sin objetivo se devuelve la
  composición observada igual, y la brecha se marca **no disponible**. Es la
  diferencia entre «no hay brecha» y «no se puede calcular la brecha», y la
  respuesta las distingue explícitamente.

Y el cruce de dos dimensiones (P1): sexo × tramo etario, que es donde
aparecen los huecos que cada dimensión por separado esconde (un panel puede
tener bien el sexo y bien la edad, y no tener ninguna mujer de más de 65).

Todo pasa en la bóveda: los atributos demográficos son autoritativos acá.
"""

from . import db, demografia
from .errores import DatosInvalidos, NoEncontrado

TOLERANCIA = 0.005
"""Cuánto se le perdona a la suma de proporciones de una dimensión. Un
universo cargado a dos decimales no suma exactamente 1 casi nunca (0,52 +
0,48 sí, pero siete tramos etarios redondeados, no), así que se acepta medio
punto porcentual de desvío. Más que eso es un error de carga."""


def _validar_objetivos(objetivos, categoricas=None):
    """Deja los objetivos en forma canónica y verifica cada dimensión.

    La validación es por dimensión, no sobre el total: cargar solo `sexo`
    tiene que poder hacerse, y entonces el total del panel es 1, no 2.

    R3.14 — `categoricas` son las dimensiones de cuota admitidas, que salen
    del catálogo de atributos. Sin ellas se valida contra el núcleo.
    """
    if not objetivos:
        raise DatosInvalidos("No hay objetivos para cargar.")
    categoricas = list(categoricas or demografia.DIMENSIONES_CATEGORICAS)

    por_dimension = {}
    for crudo in objetivos:
        if not isinstance(crudo, dict):
            raise DatosInvalidos(f"Objetivo mal formado: {crudo!r}.")
        dimension = (crudo.get("dimension") or "").strip().lower()
        if dimension not in categoricas:
            raise DatosInvalidos(
                f"Dimensión desconocida para composición: {dimension!r}.",
                {"dimensiones_validas": list(categoricas)},
            )
        categoria = str(crudo.get("categoria") or "").strip()
        if not categoria:
            raise DatosInvalidos(
                f"Falta la categoría de un objetivo de «{dimension}»."
            )
        crudo_proporcion = crudo.get("proporcion")
        if crudo_proporcion is None:
            crudo_proporcion = crudo.get("proporcion_objetivo")
        try:
            proporcion = float(crudo_proporcion)
        except (TypeError, ValueError):
            raise DatosInvalidos(
                f"Proporción inválida en {dimension}/{categoria}: "
                f"{crudo_proporcion!r}."
            )
        if not 0 <= proporcion <= 1:
            raise DatosInvalidos(
                f"La proporción de {dimension}/{categoria} tiene que estar entre "
                f"0 y 1; llegó {proporcion}. Si viene en porcentaje, dividila "
                f"por 100."
            )
        anterior = por_dimension.setdefault(dimension, {})
        if categoria in anterior:
            raise DatosInvalidos(
                f"La categoría {categoria} de {dimension} viene dos veces."
            )
        anterior[categoria] = proporcion

    for dimension, categorias in por_dimension.items():
        suma = sum(categorias.values())
        if abs(suma - 1.0) > TOLERANCIA:
            raise DatosInvalidos(
                f"Las proporciones de «{dimension}» suman {suma:.4f} y tienen "
                f"que sumar 1. Un universo de referencia incompleto haría que "
                f"todas las brechas de esa dimensión estén mal.",
                {
                    "dimension": dimension,
                    "suma": round(suma, 6),
                    "tolerancia": TOLERANCIA,
                    "categorias": categorias,
                },
            )
    return por_dimension


def guardar_objetivo(conn, panel_id, objetivos):
    """Carga (o reemplaza) el universo de referencia de un panel.

    Reemplaza **por dimensión**: cargar sexo no borra el tramo etario que ya
    estaba, pero volver a cargar sexo sí reemplaza las categorías anteriores
    de sexo. Si no fuera así, quitar una categoría del universo sería
    imposible desde la app.
    """
    if not db.una(conn, "select 1 from panel where id = %s", (panel_id,)):
        raise NoEncontrado(f"No existe el panel {panel_id}.")
    por_dimension = _validar_objetivos(objetivos, demografia.categoricas(conn))

    for dimension, categorias in por_dimension.items():
        db.ejecutar(
            conn,
            "delete from objetivo_composicion where panel_id = %s and dimension = %s",
            (panel_id, dimension),
        )
        for categoria, proporcion in categorias.items():
            db.ejecutar(
                conn,
                """
                insert into objetivo_composicion
                       (panel_id, dimension, categoria, proporcion_objetivo)
                     values (%s, %s, %s, %s)
                """,
                (panel_id, dimension, categoria, proporcion),
            )
    return obtener_objetivo(conn, panel_id)


def obtener_objetivo(conn, panel_id):
    filas = db.todas(
        conn,
        """
        select dimension, categoria, proporcion_objetivo
          from objetivo_composicion
         where panel_id = %s
         order by dimension, categoria
        """,
        (panel_id,),
    )
    por_dimension = {}
    for f in filas:
        por_dimension.setdefault(f["dimension"], {})[f["categoria"]] = float(
            f["proporcion_objetivo"]
        )
    return {
        "panel_id": panel_id,
        "dimensiones": por_dimension,
        "items": [
            {
                "dimension": f["dimension"],
                "categoria": f["categoria"],
                "proporcion_objetivo": float(f["proporcion_objetivo"]),
            }
            for f in filas
        ],
    }


def borrar_objetivo(conn, panel_id, dimension=None):
    """Saca el objetivo de una dimensión (o el del panel entero).

    Después de esto la composición de esa dimensión vuelve a ser puramente
    descriptiva, con la brecha marcada no disponible.
    """
    if dimension:
        n = db.ejecutar(
            conn,
            "delete from objetivo_composicion where panel_id = %s and dimension = %s",
            (panel_id, dimension),
        )
    else:
        n = db.ejecutar(
            conn, "delete from objetivo_composicion where panel_id = %s", (panel_id,)
        )
    return {"panel_id": panel_id, "dimension": dimension, "borradas": n}


SIN_DATO = "(sin dato)"
"""Cómo se nombra a quien no tiene valor para la dimensión.

R3.14.d — se informa **aparte**, nunca dentro de una categoría: si los sin
dato engrosaran una categoría real, la brecha de esa categoría mentiría y la
cuota se cerraría con gente que no se sabe si corresponde."""


def _observada(conn, panel_id, dimension, estado="activo"):
    """Cuántos miembros hay en cada categoría de una dimensión.

    R3.14 — la dimensión puede ser cualquier atributo del catálogo, así que
    el conteo va contra `v_atributo_persona` y no contra columnas fijas. Quien
    no tiene valor no tiene fila en la vista: cae en `(sin dato)`.
    """
    filas = db.todas(
        conn,
        """
        select coalesce(va.valor, %s) as categoria,
               count(*)::int as observados
          from membresia m
          join persona p on p.id_persona = m.id_persona
          left join v_atributo_persona va
                 on va.id_persona = p.id_persona and va.atributo = %s
         where m.panel_id = %s and (%s::text is null or m.estado = %s::text)
         group by 1
         order by 1
        """,
        (SIN_DATO, dimension, panel_id, estado, estado),
    )
    return {f["categoria"]: f["observados"] for f in filas}


def contar_miembros(conn, panel_id, estado="activo"):
    fila = db.una(
        conn,
        """
        select count(*)::int as n from membresia
         where panel_id = %s and (%s::text is null or estado = %s::text)
        """,
        (panel_id, estado, estado),
    )
    return fila["n"]


def _dimension(conn, panel_id, dimension, objetivos, total, estado):
    observada = _observada(conn, panel_id, dimension, estado)
    del_objetivo = objetivos.get(dimension) or {}
    hay_objetivo = bool(del_objetivo)

    # R3.14.d — los «sin dato» salen de la lista de categorías y se informan
    # aparte. Dejarlos adentro tenía dos efectos, los dos malos: aparecían
    # como una categoría de cuota que nadie cargó, y su peso en el
    # denominador bajaba la proporción observada de todas las demás, con lo
    # cual la brecha marcaba un déficit que no existía.
    sin_dato = observada.pop(SIN_DATO, 0)
    con_dato = sum(observada.values())

    categorias = []
    for categoria in sorted(set(observada) | set(del_objetivo)):
        observados = observada.get(categoria, 0)
        # La composición se compara contra el universo **entre quienes tienen
        # el dato**: es la única base sobre la que las proporciones suman 1 y
        # la comparación significa algo.
        proporcion = (observados / con_dato) if con_dato else 0.0
        fila = {
            "categoria": categoria,
            "observados": observados,
            "proporcion_observada": round(proporcion, 4),
            "proporcion_objetivo": None,
            "brecha": None,
            "faltan": None,
        }
        if hay_objetivo and categoria in del_objetivo:
            objetivo = del_objetivo[categoria]
            # El faltante, en cambio, se cuenta contra el panel entero: es
            # gente que hay que sumar, y el panel es del tamaño que es.
            esperados = round(objetivo * total)
            fila.update({
                "proporcion_objetivo": round(objetivo, 4),
                # Brecha en puntos de proporción: negativa = falta gente.
                "brecha": round(proporcion - objetivo, 4),
                "faltan": max(0, esperados - observados),
                "sobran": max(0, observados - esperados),
                "esperados": esperados,
            })
        categorias.append(fila)

    return {
        "dimension": dimension,
        "brecha_disponible": hay_objetivo,
        "motivo_sin_brecha": None if hay_objetivo else (
            f"No hay universo de referencia cargado para «{dimension}»: la "
            f"composición es descriptiva y la brecha no se puede calcular."
        ),
        "categorias": categorias,
        "con_dato": con_dato,
        "sin_dato": sin_dato,
        "proporcion_sin_dato": round(sin_dato / total, 4) if total else 0.0,
        "aviso_sin_dato": (
            f"{sin_dato} miembro(s) no tienen «{dimension}» cargado. No se "
            f"cuentan en ninguna categoría: las proporciones de arriba son "
            f"sobre los {con_dato} que sí lo tienen."
        ) if sin_dato else None,
        # Índice de disimilitud: la mitad de la suma de las diferencias
        # absolutas. Se lee como «qué fracción del panel habría que mover de
        # categoría para calzar con el universo». 0 = calza exacto.
        "disimilitud": round(
            sum(abs(c["brecha"]) for c in categorias if c["brecha"] is not None) / 2, 4
        ) if hay_objetivo else None,
    }


def cruce(conn, panel_id, dimension_a, dimension_b, estado="activo"):
    """P1 — composición por el cruce de dos dimensiones.

    Es donde aparecen los huecos que las marginales esconden: un panel puede
    tener la proporción correcta de mujeres y la correcta de mayores de 65, y
    no tener ninguna mujer mayor de 65.

    No lleva brecha: un objetivo cruzado es un universo distinto, con una
    celda por combinación, y `objetivo_composicion` guarda marginales. El
    cruce es descriptivo por diseño, y lo dice.
    """
    categoricas = demografia.categoricas(conn)
    for dimension in (dimension_a, dimension_b):
        if dimension not in categoricas:
            raise DatosInvalidos(
                f"Dimensión desconocida para composición: {dimension!r}.",
                {"dimensiones_validas": list(categoricas)},
            )
    if dimension_a == dimension_b:
        raise DatosInvalidos("El cruce necesita dos dimensiones distintas.")

    total = contar_miembros(conn, panel_id, estado)
    filas = db.todas(
        conn,
        """
        select coalesce(va.valor, %s) as a,
               coalesce(vb.valor, %s) as b,
               count(*)::int as observados
          from membresia m
          join persona p on p.id_persona = m.id_persona
          left join v_atributo_persona va
                 on va.id_persona = p.id_persona and va.atributo = %s
          left join v_atributo_persona vb
                 on vb.id_persona = p.id_persona and vb.atributo = %s
         where m.panel_id = %s and (%s::text is null or m.estado = %s::text)
         group by 1, 2
         order by 1, 2
        """,
        (SIN_DATO, SIN_DATO, dimension_a, dimension_b, panel_id, estado, estado),
    )
    return {
        "panel_id": panel_id,
        "dimensiones": [dimension_a, dimension_b],
        "miembros": total,
        "brecha_disponible": False,
        "motivo_sin_brecha": (
            "El universo de referencia se carga por dimensión (marginales), no "
            "por celda cruzada: el cruce es descriptivo."
        ),
        "celdas": [
            {
                dimension_a: f["a"],
                dimension_b: f["b"],
                "observados": f["observados"],
                "proporcion_observada": round(f["observados"] / total, 4) if total else 0.0,
            }
            for f in filas
        ],
    }


def composicion(conn, panel_id, dimensiones=None, estado="activo",
                cruce_de=None):
    """R2.3 — composición del panel, con brecha si hay objetivo cargado."""
    if not db.una(conn, "select 1 from panel where id = %s", (panel_id,)):
        raise NoEncontrado(f"No existe el panel {panel_id}.")

    categoricas = demografia.categoricas(conn)
    pedidas = [d.strip().lower() for d in (dimensiones or categoricas)]
    desconocidas = [d for d in pedidas if d not in categoricas]
    if desconocidas:
        raise DatosInvalidos(
            f"Dimensiones desconocidas: {desconocidas}.",
            {"dimensiones_validas": list(categoricas)},
        )

    objetivos = obtener_objetivo(conn, panel_id)["dimensiones"]
    total = contar_miembros(conn, panel_id, estado)

    salida = {
        "panel_id": panel_id,
        "estado_membresia": estado,
        "miembros": total,
        "objetivo_cargado": bool(objetivos),
        "dimensiones": [
            _dimension(conn, panel_id, dimension, objetivos, total, estado)
            for dimension in pedidas
        ],
    }
    if cruce_de:
        salida["cruce"] = cruce(conn, panel_id, cruce_de[0], cruce_de[1], estado)
    return salida
